### ISSUE TITLE
Decide on if we include the list of elements in addition to the native serialization of elements.

### ISSUE THREAD

**Issue by goneall** [spdx/spdx-3-model#505]
In the serialization meeting on 28 Sept. 2023, we made the following decision we came close to deciding on how to serialize the "where do we start" information (see issue #478 for context).  We agreed the model will have a class to represent a serializable collection that includes `rootElements` to find where you start + the list of elements that are to be serialized.

We could not reach consensus on how this class is actually serializes the list of elements and would like input from the larger SPDX tech community.

The scope of this question is specific to serialization format that natively support a closed list of elements - for example, the `@graph` in JSON-LD can be used to represent the closed list of elements used in the serialized data.

For the purposes of this issue, we will name the class "XCollection" to avoid conflating the "SPDX document" discussion with this decision.

Proposal A is to serialize the list of Element ID's as an `element` property of the `XCollection`.

For example:
```
{
  "@context": [
    "https://spdx.github.io/spdx-3-model/rdf/context.json",
    {
      "spdx-example": "http://spdx.org/spdxdocs/spdx-example-444504E0-4F89-41D3-9A0C-0305E82C3301#"
    }
  ],
  "@graph": [
    {
      "type": "Package",
      "spdxId": "spdx-example:SPDXRef-Package",
       ...
    },
    {
      "type": "File",
      "spdxId": "spdx-example:SPDXRef-CommonsLangSrc",
     ...
    },
    {
      "type": "XCollection",
      "spdxId": "spdx-example:x-collection",
      "rootElement": [
        "spdx-example:SPDXRef-File"
      ],
      "element": [
        "spdx-example:SPDXRef-Package",
        "spdx-example:SPDXRef-CommonsLangSrc"
      ]
   }
}
```

Proposal B is to not serialize the element property of the `XCollection`.

For example:

```
{
  "@context": [
    "https://spdx.github.io/spdx-3-model/rdf/context.json",
    {
      "spdx-example": "http://spdx.org/spdxdocs/spdx-example-444504E0-4F89-41D3-9A0C-0305E82C3301#"
    }
  ],
  "@graph": [
    {
      "type": "Package",
      "spdxId": "spdx-example:SPDXRef-Package",
       ...
    },
    {
      "type": "File",
      "spdxId": "spdx-example:SPDXRef-CommonsLangSrc",
     ...
    },
    {
      "type": "XCollection",
      "spdxId": "spdx-example:x-collection",
      "rootElement": [
        "spdx-example:SPDXRef-File"
      ],
     }
   ]
}
```

From the serialization call minutes:

* Advantage of the serlializing the element list [Proposal A] is that you could use it to verify the serialized elements - similar to a packing list in a box of content or a manifest
* Advantage of not serializing the list [Proposal B] is brevity and not duplicating information - note: if the information is used for verification it really isn't duplicated
* Decision - if we did always serialize the element list, it would be primarily used for verification and the the native serialization contains is what is really contained in the serialization (as captured in the "SerializedCollection" which represents SPDX data which has already been serialized) - we would document that the element list is provided for verification purposes and not to be confused with the list of actual elements serialized. The "SerializableCollection" [a.k.a. "XCollection"] and "SerializedCollection" classes work in concert to express what 'should' but in the serialized content and what 'is' in the serialized content and are machine comparable to provide clear verification.

**Comment by goneall** [comment]
@sbarnum @nishakm @davaya @maxhbr Please review the above description of the decision we would like to tee up for Tuesday's call.  I copied / pasted from the minutes with a minimal amount of modification.  I'll send out an email to the tech distribution list sometime this weekend to ask for review before Tuesday's call.

**Comment by davaya** [comment]
> We agreed the model will have a class to represent a serializable collection that includes rootElements to find where you start + the list of elements that are to be serialized.

I agree.  Where-to-start is independent of serialization format and therefore it should be in a serializable collection *when it is helpful*.

But in most serialized data it is not helpful:
* non-collections (License, Actor, Artifact and other elements that are not ElementCollections)
* a single ElementCollection (Bundle, Bom, Sbom) and all elements listed in that collection - start with the ElementCollection.
* non-nested ElementCollections (two or more Bundle, Bom, Sbom elements that do not contain each other.  Two or more collections may contain the same non-collection elements without problem.) - start with each ElementCollection, order does not matter.

Where-to-start is helpful only in the case where serialized data includes an ElementCollection that includes another ElementCollection.  It indicates the outer-most ElementCollections, allowing the consumer to begin some processing before determining the roots of the ElementCollection graph.   But at some point consumers must process the graph even if where-to-start is not included in the serialized data or is incorrect.

> Advantage of the serlializing the element list [Proposal A] is that you could use it to verify the serialized elements - similar to a packing list in a box of content or a manifest

As noted in #478, Proposal A is never useful in verifying serialized elements.  Unlike a shipping box where items in the box can go missing and a manifest can verify the shipping contents, serialized data is an atomic unit.
* If the consumer has the data serialized by the producer, the consumer has exactly every element serialized by the producer
* If the consumer has data different from that serialized by the producer, a manifest included in that data could be different from the producer's manifest.
Thus a list of elements separate from the elements themselves is both redundant and not useful for verification.

**Comment by goneall** [comment]
> I agree. Where-to-start is independent of serialization format and therefore it should be in a serializable collection when it is helpful.

@davaya I recall on one of our calls we agreed that if you did not include the `XCollection` with the `rootElement` then all elements would be considered "root elements".  So, for the purposes of this issue - let's assume this is a collection where the `rootElement` is helpful.

**Comment by maxhbr** [comment]
I would prefer **B**, although the semantics of `Collection` do not work with it. I think for variant **B** it can not be a Collection and that the property pointing to the "described elements" should not be `rootElement`, which is a purely technical property of Collections.

If we can not distance from the Collection I think it needs to be **A**.

**Comment by goneall** [comment]
@maxhbr - I think B can still represent the semantics of Collection - we are just specifying that in the serialization / deserialization the collection elements are represented in the native serialization format.

**Comment by maxhbr** [comment]
Ok, you know my opinion on serialization rules for JSON-LD :wink:

**Comment by davaya** [comment]
@maxhbr 
> If we can not distance from the Collection I think it needs to be A.

Later in the call we discussed names of the elements, which we will pick up on Thursday.  But there always needs to be a set of elements independent of serialization format, plus a set of elements in an actual document (specific to the serialization format of that document).

When we were discussing producer-centric (A) vs. consumer-centric (B), B referred to the element that describes a payload.  In this issue, A is "set of elements plus a list-of-elements property" and B is "set of elements without a separate list-of-elements property".  For 3.0 we decided on B - serialize the elements but not a separate list-of-elements.

That still leaves the consumer-centric use case, which I guess we'll now have to call C, and select names to the B and C classes.
